docs/adr/adr-013-blocksync-overhaul-part-2.md

```diff
@@ -0,0 +1,418 @@
+the network. It is slower than block production for the EDS sizes __>128__, but has higher retrieval guarantees during
+data withholding attack. (link)
+
+- __FN/LN/BNs leverage implemented discovery mechanism outlined in ADR-008(TODO Link).__ The discovery in a node maintains connections
+with FN-subnetwork. This ensures the node will receive new notifications from `ShrEx/Sub` and can request data via
+`ShrEx/EDS` and `ShrEx/ND`.
+
+- __FN/BNs notify connected LN/FNs when they are ready to serve a recent EDS using via [`ShrEx/Sub`](#ShrEx/Sub Protocol).__
+This lazy-push approach provides reliable EDS dissemination properties and turns out to be simpler than technics similar
+to Bitswap's wantlist.(TODO link)
+
+## Protocols Specs
+
+### ShrEx/EDS Protocol
+
+ShrEx/EDS is a pull-based protocol with client-server model, where LNs are clients, BNs are servers
+and FNs are both. The protocol has only one request-response interaction followed by EDS stream.
+
+The protocol is designed over plain libp2p streams with `/shrex/eds/0.0.1` as the protocol ID.
+
+#### Protobuf Schema
+
+##### Request
+
+```protobuf
+syntax = "proto3";
+
+message EDSRequest {
+  bytes hash = 1; // identifies the requested EDS.
+}
+```
+
+##### Response
+
+```protobuf
+syntax = "proto3";
+
+enum Status {
+  OK = 100; // data found
+  NOT_FOUND = 200; // data not found
+  REFUSED = 201; // request refused
+}
+
+message EDSResponse {
+  Status status = 1;
+}
+```
+
+#### Streaming
+
+After flushing `EDSResponse`, server starts streaming EDS file to client over the libp2p stream.
+
+#### Backpressure
+
+The streaming part of the protocol does not provide any mechanics for backpressure and offloads them to layer below.
+Client-side buffer with reasonable default size may still be added.
+
+[Backpressure][backpressure-quic] is inherited from the QUIC. The underlying default reliable transport and stream
+multiplexing protocol used in `celestia-node`.
+
+The common TCP transport and yamux multiplexer combination, does not provide reliable backpressure on stream level,
+unlike QUIC. Therefore, using TCP+yamux is not recommended and may lead to unexpected stream and connections resets in case of highly
+congested links.
+
+#### Flow
+
+// TODO Make a simple sequence diagram
+
+### ShrEx/ND Protocol
+
+ShrEx/ND is a pull-based protocol with client-server model, where LNs are clients and BNs/FNs are servers.
+The protocol has only one request-response interaction to get all the data/shares addressed by namespace
+
+The protocol is designed over plain libp2p streams with `/shrex/eds/0.0.1` as the protocol ID.
+
+#### Protobuf Schema
+
+##### Request
+
+```protobuf
+syntax = "proto3";
+
+message NDRequest {
+  bytes hash = 1; // identifies the EDS to get ND from
+  bytes namespace = 2; // namespace of ND
+}
+```
+
+##### Response
+
+```protobuf
+syntax = "proto3";
+
+enum Status {
+  OK = 100; // data found
+  NOT_FOUND = 200; // data not found
+  REFUSED = 201; // request refused
```

Comment:
confusing, why not

OK 0
NOT_FOUND 1
REFUSED 2

Reply:
I followed the HTTP codes pattern. Don't have a strong opinion on this